src/models/Model.php

```diff
@@ -44,45 +47,127 @@ public function validate()
-                $getter = 'get' . ucfirst($property);
-                if (method_exists($this, $getter)) {
-                    $value = $this->$getter();
+                $getter = 'get' . ucfirst((string) $property);
+
+                $value = method_exists($this, $getter) ? $this->$getter() : $this->$property ?? null;
+
+                if ($validator === 'required') {
+                    if ($value === null || $value === '' || ($value === [])) {
+                        $errors[$property][] = 'This field is required.';
+                    }
+                    continue;
+                }
+
+                if (is_callable($validator)) {
+                    $result = call_user_func($validator, $value, $this);
+
+                    if ($result !== true) {
+                        $errors[$property][] = is_string($result) ? $result : "Validation failed for $property.";
+                    }
                 } else {
-                    $value = isset($this->$property) ? $this->$property : null;
+                    // Skip validation for string/integer/float validators if value is null (unless marked as required)
+                    if ($value === null && in_array($validator, ['string', 'integer', 'float', 'email'])) {
+                        continue;
+                    }
+
+                    switch ($validator) {
+                        case 'string':
+                            if (!is_string($value)) {
+                                $errors[$property][] = 'Must be a string.';
+                            }
+                            break;
+                        case 'integer':
+                            if (!is_int($value)) {
+                                $errors[$property][] = 'Must be an integer.';
+                            }
+                            break;
+                        case 'float':
+                            if (!is_float($value) && !is_int($value)) {
+                                $errors[$property][] = 'Must be a float.';
+                            }
+                            break;
+                        case 'email':
+                            if (!is_string($value) || !filter_var($value, FILTER_VALIDATE_EMAIL)) {
+                                $errors[$property][] = 'Must be a valid email address.';
+                            }
+                            break;
+                        default:
+                            $errors[$property][] = "Unknown validator: $validator";
+                    }
                 }
+            }
+        }
+
+        return $errors === [] ? true : $errors;
+    }
+
+    /**
+     * Validates model properties based on rules(), excluding the hash field.
+     * This is useful for validating before hash generation.
+     * Returns true if all validations pass, otherwise returns an array of error messages.
+     *
+     * @return true|array
+     */
+    public function validateExceptHash()
```

Comment:
This method does not comply with DRY principle. If we need to skip validation of one attribute, in one model, it must be done in that model, but not in the main one. This must be refactored.